File: src/main/java/novemberizing/rx/schedulers/Local.java
package novemberizing.rx.schedulers;

import novemberizing.ds.ConditionalList;
import novemberizing.ds.Executable;
import novemberizing.ds.Factory;
import novemberizing.util.Log;

/**
 *
 * @author novemberizing, me@novemberizing.net
 * @since 2017. 1. 17.
 */
@SuppressWarnings("unused")
public class Local extends Scheduler {
    private static final String Tag = "novemberizing.rx.schedulers.local";

    private static Factory<novemberizing.rx.Scheduler> __factory = null;
    private static ThreadLocal<novemberizing.rx.Scheduler> __schedulers = new ThreadLocal<>();

    public static void Set(Factory<novemberizing.rx.Scheduler> factory){ __factory = factory; }

    public static novemberizing.rx.Scheduler Get(){
        Log.f(Tag, "");
        novemberizing.rx.Scheduler scheduler = __schedulers.get();
        if(scheduler==null){
            scheduler = __factory!=null ? __factory.create() : new Local();
            if(scheduler==null){
                scheduler = new Local();
            }
            __schedulers.set(scheduler);
        }
        return scheduler;
    }

    public Local(){
        Log.f(Tag, "");
        __q = new ConditionalList<>();
    }

    @Override
    public void executed(Executable executable) {
        Log.f(Tag, "");
        super.executed(executable);
        onecycle();
    }

    @Override
    public void completed(Executable executable) {
        Log.f(Tag, "");
        super.completed(executable);
        onecycle();
    }

    @Override
    public void dispatch(Executable executable) {
        Log.f(Tag, "");
        super.dispatch(executable);
        onecycle();
    }

    @Override
    public void dispatch(Executable executable, Executable... executables) {
        Log.f(Tag, "");
        super.dispatch(executable, executables);
        onecycle();
    }

    @Override
    public void dispatch(Executable[] executables) {
        Log.f(Tag, "");
        super.dispatch(executables);
        onecycle();
    }
}

File: src/main/c++/novemberizing/rx/schedulers/local.cc
#include "local.hh"

namespace novemberizing { namespace rx { namespace schedulers {

Thread::Local<Scheduler *> Local::__schedulers;

Scheduler * Local::Get(void)
{
    Scheduler * scheduler = __schedulers.get();
    if(scheduler==nullptr)
    {
        __schedulers.set(scheduler=new Local());
    }
    return scheduler;
}

Local::Local(void)
{
    FUNCTION_START("");
    FUNCTION_END("");
}

Local::~Local(void)
{
    FUNCTION_START("");
    type::size remain = 0;
    do {
        __q.lock();
        while(__q.size()>0){
            __q.pop([this](Executable * executable){
                if(executable!=nullptr)
                {
                    __q.unlock();
                    synchronized(&__executables, __executables.add(executable));
                    executable->execute(this);
                    __q.lock();
                }
                else
                {
                    NOTICE_LOG("executable==nullptr");
                }
            });
        }
        synchronized(&__executables, remain = __executables.size());
        __q.unlock();
    } while(remain>0);
    synchronized(&__deletes,{
        while(__deletes.size()>0)
        {
            __deletes.pop([](Executable * executable){
                if(executable!=nullptr) {
                    delete executable;
                }
            });
        }
    });
    FUNCTION_END("");
}

int Local::run(void)
{
    while(!cancel()){
        __q.lock();
        type::size current = __q.size();
        for(type::size i = 0;i<current && __q.size()>0;i++)
        {
            __q.pop([this](Executable * executable){
                if(executable!=nullptr)
                {
                    __q.unlock();
                    synchronized(&__executables, __executables.add(executable));
                    executable->execute(this);
                    __q.lock();
                }
                else
                {
                    NOTICE_LOG("executable==nullptr");
                }
            });
        }
        __q.unlock();
        synchronized(&__deletes,{
            while(__deletes.size()>0)
            {
                __deletes.pop([](Executable * executable){
                    if(executable!=nullptr) {
                        delete executable;
                    }
                });
            }
        });
    }
    return Success;
}

void Local::dispatch(Executable * executable)
{
    FUNCTION_START("");
    synchronized(&__q,{
        if(executable!=nullptr){ __q.push(executable); }
        __q.resume();
    });
    FUNCTION_END("");
}

void Local::dispatch(std::initializer_list<Executable *> executables)
{
    FUNCTION_START("");
    synchronized(&__q,{
        for(std::initializer_list<Executable *>::iterator it = executables.begin();executables.end();it++)
        {
            if(*it!=nullptr){ __q.push(*it); }
        }
        __q.resume(true);
    });
    FUNCTION_END("");
}

void Local::completed(Executable * executable)
{
    FUNCTION_START("");
    if(executable!=nullptr)
    {
        synchronized(&__executables, __executables.del(executable));
    }
    FUNCTION_END("");
}

void Local::executed(Executable * executable)
{
    FUNCTION_START("");
    if(executable!=nullptr)
    {
        synchronized(&__executables, __executables.del(executable));
        synchronized(&__q, {
            __q.front(executable);
            __q.resume();
        });
    }
    FUNCTION_END("");
}

} } }
